tests/TestCase/View/Helper/FormHelperTest.php

```diff
@@ -8496,4 +8496,86 @@ public function testNestedLabelInput()
         ];
         $this->assertHtml($expected, $result);
     }
+
+    /**
+     * testRadioAttributes method
+     *
+     * Test generation of radio widget with additional attributes.
+     *
+     * @return void
+     */
+    public function testRadioAttributes()
+    {
+        $result = $this->Form->radio('Model.field', ['option A'], ['class' => 'my-class', 'data-ref' => 'custom-attr']);
+        $expected = [
+            'input' => ['type' => 'hidden', 'name' => 'Model[field]', 'value' => ''],
+            'label' => ['for' => 'model-field-0'],
+            ['input' => [
+                'type' => 'radio', 'name' => 'Model[field]',
+                'value' => '0', 'id' => 'model-field-0',
+                'class' => 'my-class', 'data-ref' => 'custom-attr'
+            ]],
+            'option A',
+            '/label'
+        ];
+        $this->assertHtml($expected, $result);
+    }
+
+    /**
+     * testCheckboxAttributes method
+     *
+     * Test generation of checkbox widget with additional attributes.
+     *
+     * @return void
+     */
+    public function testCheckboxAttributes()
+    {
+        $result = $this->Form->checkbox('Model.field', ['class' => 'my-class', 'data-ref' => 'custom-attr']);
```

Comment:
I'd almost prefer these tests to exist at the widget level. While the public interface is FormHelper it is useful to capture these kinds of attribute permutations in the widget tests themselves, as the FormHelper test is a beast as you're well aware.